src/Disks/S3/registerDiskS3.cpp

```diff
@@ -178,12 +178,13 @@ void registerDiskS3(DiskFactory & factory)
 
         String metadata_path = config.getString(config_prefix + ".metadata_path", context->getPath() + "disks/" + name + "/");
         fs::create_directories(metadata_path);
+        auto metadata_disk = std::make_shared<DiskLocal>(name + "-metadata", metadata_path, 0);
```

Comment:
So still in this PR only `DiskLocal` is used. Are you planning to add implementation with disks `s3` / `hdfs` ? 

With current set up it will be very tricky since `DisksS3` and `DiskHDFS` are inherited from `IDiskRemote` and they are now passed there as DiskPtr's to reimplement most methods, but the same methods will have to be overriden by them to make passed `DiskPtr` work as expected, but then methods of `IDiskRemote` will not be called.

So just passing `DiskPtr` of the needed disk type is far not enough.

Reply:
DisksS3 and DiskHDFS are inherited from IDiskRemote, and most metadata of s3 / hdfs are access(read/write) in methods of IDiskRemote, and most methods of IDiskRemote just implement the logical of DiskLocal again. And the metadata can only store in Local, there is little extendiability if we want store metadata into shard storage.
If there is fast enough distributed file system, but not s3 / hdfs, it is easy to replace DiskLocal as metadata store after this PR only if the DFS implement interfaces of IDisk.
This PR not change the architecture of DiskS3 and IDiskRemote, but can make it more flexible to replace metadata store

Reply:
> If there is fast enough distributed file system, but not s3 / hdfs, it is easy to replace DiskLocal as metadata store after this PR only if the DFS implement interfaces of IDisk.

But why not s3 / hdfs, mainly s3? They are the only options we have now.
So it makes sense to allow storing metadata there.

> This PR not change the architecture of DiskS3 and IDiskRemote, but can make it more flexible to replace metadata store

My point was that it is impossible with these changes to pass `S3Disk`/`HDFSDisk` as DiskPtr instead of `DiskLocal`, and now in this PR only `DiskLocal` is passed.
Yes, adding DiskPtr as a parameter makes it more flexible, but will not allow store metadata on s3 (because of what I write in my first comment), but it is what we need.

Reply:
> If there is fast enough distributed file system, but not s3 / hdfs, it is easy to replace DiskLocal as metadata store after this PR only if the DFS implement interfaces of IDisk.

So if it was not your intention to allow storing metadata in s3, but in some possibly implemented in the future fs, then it would be ok, but if it did not conflict with the possibility to allow storing metadata in s3. Because when we will want to do that, we will have to change this implementation because it does not allow to pass `DiskS3` as DiskPtr into `IDiskRemote` the same way as you now pass `DiskLocal`.

Reply:
Metadata store local disk rather than s3, because of long latency and poor performance of rename.  So we'd better to store metadata and data separately. HDFS's metadata performance is also not good enough and lark of hard link.

Reply:
It makes sense but high chances are that we will want to allow storing metadata in s3 also.

But ok, these changes are good, though I thought you wanted to allow storing metadata in s3/hdfs, now it only removes copy paste from DiskLocal.

Ok, lets wait for tests..

Reply:
> Metadata store local disk rather than s3, because of long latency and poor performance of rename. 

I'm not sure how these impact metadata storage, as it moves much slower than data itself (which is stored on S3 just fine).
I.e. it just makes things worse, not better